sub/sd_ass.c

```diff
@@ -294,8 +294,19 @@ static bool is_animated(char *s)
     bool valid_event = false;
     bool valid_tag = false;
     while (*s) {
-        if (!in_tag && s[0] == '{')
+        if (!in_tag) {
+            s = strchr(s, '{');
```

Comment:
Added a reverse search from the `{` to count `\`s and determine if the `{` is escaped.

Reply:
> Added a reverse search from the `{` to count `\`s and determine if the `{` is escaped.

AFAIK that's not how it works. I think `\` is only escaping in `\n`, `\N`, `\h`, `\{`. Specifically, `\\` is not an escaped backslash.

The whole point of https://github.com/libass/libass/issues/194 is that it's impossible to output a backslash followed by a tag (or by n, N, or h), like so `xxx\\{whatever}`, because it's interpreted as literal `xxx\` folloiwed by a literal (escaped) `{`, and I don't think it changed in the last decade.

So the rule is, AFAIK: "`\{` is always escaped, regardless of how many `\` may or may not appear porior to it".

> This is much faster than looping byte-wise, since the libc generally has
vectorized implementations of these routines.

While I assume this statement is true, are there real world use cases where the speed difference by this change is meaningful or even measurable?

Reply:
@avih 
> Specifically, `\\` is not an escaped backslash.

Ah, you're right, I'd misread the escape handling code. Re-pushed with a fix (which also simplifies the check substantially).

> are there real world use cases where the speed difference by this change is meaningful or even measurable?

I ran into this substantial CPU usage in `is_animated` while profiling [08_TS_joletb_preasswipe.ass.zip](https://github.com/libass/libass/files/15445460/08_TS_joletb_preasswipe.ass.zip); it took up nearly 600 megacycles, 15% of the total over my test period. This PR reduces that overhead by a factor of >20.

@na-na-hi 
> After the reverse search is added, does the claimed performance benefit still exist, and how much?

The reverse search is now removed (as it turned out not to be necessary), but when it existed, it did not have a significant impact on performance (since real-world cases of long strings of `\`s immediately before `{`s are essentially nonexistent).

Reply:
> > Ah, you're right, I'd misread the escape handling code. Re-pushed with a fix

Right.

I didn't try to assess whether the original code is correct, but does the current PR change the behavior with regards to escape (or anything else)? if yes, it might be worth splitting it into functional/correctness change commit, and performance change commit, IMO.

> I ran into this substantial CPU usage in `is_animated` while profiling [08_TS_joletb_preasswipe.ass.zip](https://github.com/libass/libass/files/15445460/08_TS_joletb_preasswipe.ass.zip); it took up nearly 600 megacycles, 15% of the total over my test period. This PR reduces that overhead by a factor of >20.

This is a 80 MB ass file, which I'd think is quite uncommon, isn't it? but even with this ass file, is there a scenario with mpv where this PR makes a difference noticeable by the user?

Reply:
Very large ASS files, usually with most of the size concentrated over short periods of time, are fairly routine. This file was sent to me by its author because it had performance issues (along with a smaller version that performed worse due to cache issues); I've also opened anither PR against libass to fix a similar (though larger) issue there (another inefficient loop searching for a curly brace instead of using strchr[nul]). In a file that's already borderline on lag, saving 15% of aggregate CPU time can easily be the difference between playing smoothly and dropping frames.

I *believe* the existing code was correct escape-wise.

Reply:
> I _believe_ the existing code was correct escape-wise.

Well.. hard for me to tell, to be honest. The comment of this function states "Check if subtitle has events that would cause it to be animated inside {}", but (before this PR):
- The test `if (!in_tag && s[0] == '{') in_tag = true;` does not seem to test whether it follows a `\`, I _think_ `\{` would still end up with `in_tag = true`.
- The whole `if (s[0] == '\\') { ...}` code is done regardless of whether it's inside or outside a tag, as far as I can tell.

Dunno, the original code looks weird to me, to be honest, BUT this is not a review of the original code, and I could be totally wrong here.

Whatever.

Reply:
It doesn't matter if `is_animated` has some false positives because it only makes a difference with an image instead of a video, where it makes mpv redraw continously instead of only when the subtitle line changes, and no one uses ASS subtitles with cover art.

Reply:
> It doesn't matter if `is_animated` has some false positives

I believe you, but no one could know that by reading the code and the comments.

Reply:
The `if (s[0] == '\\') {` block always advances `s` by an additional character, so it'll consume a `\{`. That said, I think this would be substantially clearer if it used a nested loop for the within-`{}` case.

Reply:
> advances `s` by an additional character,

Right. It does advance s twice in that case. I missed that. Thanks.

> I think this would be substantially clearer if it used a nested loop for the within-`{}` case.

Well, assuming it is only interested in the "within-`{}`" case (at least judging by the comment just above is_animated, but not by the actual code), the whole loop could be only inside tags, logically similar to this:
```c
    while ((s = strchr(s, '{')) && (s == start || s[-1] != '\\')) {
        ++s;  // inside a tag

        char *end = strchr(s, '}');
        if (!end)
            /* whatever */;

        while ((s = memchr(s, '\\', end - s)) {
            ++s;  // backslash<s>
            switch (*s) {
                /* whatever */
            }
        }

        s = end + 1;
    }
```

Reply:
Yeah, that'd work. I think it's out-of-scope for this PR, though.

Reply:
@rcombs this PR does change the functionality of `is_animated`, so at the very least it's worth mentioning that (and how it changes) at the commit message.

For instance, the string `\k {}` is considered animated without this PR, but not animated with this PR.

That's because the PR skips the content till the next `{`, while the original code also "finds" things outside of tags.

As I said, hard to tell what the original intent is, because the comment and the actual code disagree on what it does.

Reply:
The new behavior is intended; noted in the commit message.

Reply:
> noted in the commit message.

Thanks.

I think this was an opportunity to rewrite it much nicer, but anyway, it's better now.

Reply:
I just didn't consider the possibility of something like `\k` preceding the `{}`. No great mystery there. Feel free to rewrite the whole thing if you'd prefer.